### ISSUE TITLE
[Bug]: SiLU Approximation Issue

### ISSUE THREAD

**Issue by wilyub** [secretflow/spu#555]
### Issue Type

Performance

### Modules Involved

MPC protocol

### Have you reproduced the bug with SPU HEAD?

Yes

### Have you searched existing issues?

Yes

### SPU Version

0.7.0b0

### OS Platform and Distribution

Linux Ubuntu 22.04

### Python Version

3.11

### Current Behavior?

Hi All. I'm trying to test out the SiLU polynomial approximation from the PUMA paper. However, when I replace the SiLU in Llama-7B with the polynomial approximation, it seems to break the transformer. I have tested out the softmax approximation from the paper and that seems to work fine. Any help on figuring out where I've messed up the implementation would be appreciated.

I have attached the code for the SiLU approximation and the LlamaMLP class I changed. For testing, I check the perplexity on the wikitext-2-raw-v1 dataset. For the softmax approximation I have a perplexity of 5.0394 (same as the original Llama), but once I add the SiLU approximation (alongside the softmax approximation) I get a "nan" perplexity.

### Standalone code to reproduce the issue

```Python
def hacked_silu(x):
    b0 = x < -8.0
    b1 = x < -4.0
    b2 = x > 4.0
    b3 = b1 ^ b2 ^ True  # x in [-4.0, 4.0)
    b4 = b0 ^ b1  # x in [-8.0, -4.0)
    # seg1 =  a[2] * x^2 + a[1] * x + a[0]
    # seg2 = b[6] * x^6 + b[4] * x^4 + b[2] * x^2 + b[0]
    a_coeffs = [-0.3067541139982155, -0.0819767021525476, -0.0055465625580307]
    b_coeffs = [
            0.0085064025895951,
            0.5,
            0.2281430841728270,
            -0.011113046708173,
            0.0002743776353465,
        ]
    x2 = torch.square(x)
    x4 = torch.square(x2)
    x6 = x2 * x4
    seg1 = a_coeffs[2] * x2 + a_coeffs[1] * x + a_coeffs[0]
    seg2 = (
        b_coeffs[4] * x6
        + b_coeffs[3] * x4
        + b_coeffs[2] * x2
        + b_coeffs[1] * x
        + b_coeffs[0]
    )
    ret = b2 * x + b4 * seg1 + b3 * seg2
    return ret


class CustomLlamaMLP(LlamaMLP):
    def __init__(self, config):
        super().__init__(config)
        self.config = config
        self.hidden_size = config.hidden_size
        self.intermediate_size = config.intermediate_size
        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
        #self.act_fn = ACT2FN[config.hidden_act]

    def forward(self, x):
        if self.config.pretraining_tp > 1:
            slice = self.intermediate_size // self.config.pretraining_tp
            gate_proj_slices = self.gate_proj.weight.split(slice, dim=0)
            up_proj_slices = self.up_proj.weight.split(slice, dim=0)
            down_proj_slices = self.down_proj.weight.split(slice, dim=1)

            gate_proj = torch.cat(
                [F.linear(x, gate_proj_slices[i]) for i in range(self.config.pretraining_tp)], dim=-1
            )
            up_proj = torch.cat([F.linear(x, up_proj_slices[i]) for i in range(self.config.pretraining_tp)], dim=-1)

            intermediate_states = (hacked_silu(gate_proj) * up_proj).split(slice, dim=2)
            down_proj = [
                F.linear(intermediate_states[i], down_proj_slices[i]) for i in range(self.config.pretraining_tp)
            ]
            down_proj = sum(down_proj)
        else:
            down_proj = self.down_proj(hacked_silu(self.gate_proj(x)) * self.up_proj(x))

        return down_proj
```

**Comment by anakinxc** [comment]
@Ye-D

**Comment by fionser** [comment]
Possible overflow in `torch.square` in the polynomial evaluation. 
It depends on the ring width (e.g., FM64, FM128) and the fixed-point precision you have used. 
May you try to add a `range clamp` before the `hacked_silu` call, for instance clamp to `[-20, 20]` (or larger?)
